### ISSUE TITLE
document module_utils.*

### ISSUE THREAD

**Issue by kontrafiktion** [ansible/ansible#16888]
##### ISSUE TYPE
- Documentation Report
##### COMPONENT NAME

internal: module_utils.*
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file =
  configured module search path = Default w/o overrides
```
##### SUMMARY

Module developers/maintainers have a hard time using module_utils.\* because there is nearly no documentation. At least I couldn't find any.

Proposal:
1. Create documentation for module_utils.*
2. Publish it on the Ansible web site
###### Details

For example, the fetch_url method does not document its return values. You have to debug and know the internally used urllib2 to understand what the "info" return value contains. 

When creating a new module, using the "requests" library, the build says: 

```
8.60s$ ansible-validate-modules --exclude 'utilities/' .
============================================================================
./cloud/digital_ocean/digital_ocean_tag.py
============================================================================
ERROR: requests import found, should use ansible.module_utils.urls instead
```

Therefore: when asking (demanding) to use a library it should be documented (as thoroughly as possible). Otherwise the entry hurdle becomes rather high

**Comment by bcoca** [comment]
We had just discussed this at the contributor summit last week.

**Comment by kontrafiktion** [comment]
See for a very early version: https://github.com/ansible/ansible/pull/16890 (superceded by #16987)

**Comment by kbreit** [comment]
Is there still effort around this? I'm new to Ansible and developing modules and I've found the lack of documentation around module_utils to be challenging. I'd be happy to help on this, but probably don't have time to take it all on myself.

**Comment by danihodovic** [comment]
I wish the `ansible.module_utils.urls` provided a  simple way to log requests / responses at HTTP level.

**Comment by acozine** [comment]
@kontrafiktion thanks for opening this issue and #16987!

Since then we've expanded and improved the documentation on `module_utils` - the latest changes came from PR #55931. 

@kontrafiktion @kbreit could you take a look at the [latest version of this page](https://docs.ansible.com/ansible/devel/dev_guide/developing_module_utilities.html) on the `devel` docs and let me know if we can close this issue? The content will be published to the `latest` docs when Ansible 2.8 is released later this month.

**Comment by samccann** [comment]
Documentation has been updated (see prior comment). If more is missing, please open individual issues/PRs to address what is lacking. Thanks!
